Add 3 basic classes for HTTP2 based NetworkClient implementation.

zzmao opened the pull request:
Implemented Http2ChannelPoolMap, MultiplexedChannelRecord and Http2MultiplexedChannelPool.
Moved HTTP/2 related classes to com.github.ambry.network.http2
Moved SSLFactory implementations to com.github.ambry.commons

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/Http2ChannelInitializer.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * A {@link ChannelInitializer} to be used with {@link Http2ChannelPoolMap}.
+ */
+public class Http2ChannelInitializer extends AbstractChannelPoolHandler {
```
Could you call this `Http2ClientChannelInitializer` and replace the one used for `Http2BlockingChannel` with this one?

zzmao replied:
Http2ClientChannelInitializer is `Channelnitializer`, which is not a `AbstractChannelPoolHandler`.
We probably need to keep two classes. 
I can rename this(Http2ChannelInitializer) to `Http2ChannelPoolHanlder` and try to avoid dup code.

zzmao replied:
Http2ClientChannelInitializer class is removed. Do inline initialize in Http2BlockingChannel::channel().

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/Http2ChannelInitializer.java:
```diff
@@ -0,0 +1,63 @@
+  @Override
+  public void channelCreated(Channel ch) throws Exception {
+    ChannelPipeline pipeline = ch.pipeline();
+    SSLFactory sslFactory = new NettySslHttp2Factory(sslConfig);
```
You should only need one shared sslFactory for all channels. I would prefer only parsing the configs once

zzmao replied:
ok

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/Http2ClientChannelInitializer.java:
```diff
@@ -26,8 +26,7 @@
 /**
- * A {@link ChannelInitializer} to be used with {@link Http2BlockingChannel}. Calling {@link #initChannel(SocketChannel)}
- * adds the necessary handlers to a channel's pipeline so that it may handle requests.
+ * A {@link ChannelInitializer} to be used with {@link Http2ChannelPoolMap}.
```
looks like this is not used with `Http2ChannelPoolMap`

zzmao replied:
fixed. For `Http2BlockingChannel`

zzmao replied:
This class is removed.

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/Http2ChannelPoolMap.java:
```diff
@@ -0,0 +1,46 @@
+  private final EventLoopGroup eventLoopGroup;
+  private final SSLConfig sslConfig;
+
+  public Http2ChannelPoolMap(SSLConfig sslConfig, EventLoopGroup eventLoopGroup) {
```
Maybe construct `SSLFactory` outside (i.e. in the `NetworkClientFactory` impl) and pass it into this constructor since it should be able to shared across all instances.

zzmao replied:
SSLFactory is used.

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/Http2ChannelPoolMap.java:
```diff
@@ -0,0 +1,46 @@
+import io.netty.channel.socket.nio.NioSocketChannel;
+
+
+public class Http2ChannelPoolMap extends AbstractChannelPoolMap<RequestInfo, ChannelPool> {
```
It seems that RequestInfo isn't exactly the right key to use for `Http2ChannelPoolMap` and `Http2ChannelInitializer` for a couple reasons:
1. It doesn't have hashcode and equals defined, and a separate object is created for each request, so a new pool will created for each request, instead of each host:port combo
2. It includes info about the request body, which is not relevant to the pool (i think).

Instead, you could use `InetSocketAddress` and pass this directly into `Bootstrap::remoteAddress`. The impl of `Bootstrap::remoteAddress(String inetHost, int inetPort)` would call `InetSocketAddress.createUnresolved(inetHost, inetPort)` anyways.

zzmao replied:
InetSocketAddress is used.

zzmao commented:
Moved http2 realted classes to `com.github.ambry.network.http2`.

This also requires to move `NettyHttp2SslFactory` to `com.github.ambry.commons`. Http2 needs `NettyHttp2SslFactory`.

So I moved both `NettyHttp2SslFactory` and `NettySslFactory` to `com.github.ambry.commons` and ambry-commons now also depends on netty-all. (if you have any suggestion, let me know.)

cgtz commented on ambry-server/src/integration-test/java/com.github.ambry.server/ServerTestUtil.java:
```diff
@@ -122,6 +124,7 @@
 import java.util.concurrent.atomic.AtomicReference;
 import javax.net.ssl.SSLSocketFactory;
 import org.junit.Assert;
+import org.junit.Test;
```
Is this import needed?

zzmao commented:
Addressed @cgtz 's comments.

zzmao commented:
@cgtz  @justinlin-linkedin   The build finally passed. Ready to merge. Thanks!
